lib/Jsrt/JsrtDiag.cpp

```diff
@@ -0,0 +1,688 @@
+    _In_ JsRuntimeHandle runtimeHandle)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        VALIDATE_INCOMING_RUNTIME_HANDLE(runtimeHandle);
+
+        JsrtRuntime * runtime = JsrtRuntime::FromHandle(runtimeHandle);
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            jsrtDebugManager->EnableAsyncBreak(scriptContext);
+        }
+
+        return JsNoError;
+    });
+}
+
+CHAKRA_API JsDiagGetBreakpoints(
+    _Out_ JsValueRef *breakpoints)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        PARAM_NOT_NULL(breakpoints);
+
+        *breakpoints = JS_INVALID_REFERENCE;
+
+        JsrtContext *currentContext = JsrtContext::GetCurrent();
+
+        Js::JavascriptArray* bpsArray = currentContext->GetScriptContext()->GetLibrary()->CreateArray();
+
+        JsrtRuntime * runtime = currentContext->GetRuntime();
+
+        ThreadContextScope scope(runtime->GetThreadContext());
+
+        if (!scope.IsValid())
+        {
+            return JsErrorWrongThread;
+        }
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        VALIDATE_IS_DEBUGGING(jsrtDebugManager);
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            jsrtDebugManager->GetBreakpoints(&bpsArray, scriptContext);
+        }
+
+        *breakpoints = bpsArray;
+
+        return JsNoError;
+    });
+}
+
+CHAKRA_API JsDiagSetBreakpoint(
+    _In_ unsigned int scriptId,
+    _In_ unsigned int lineNumber,
+    _In_ unsigned int columnNumber,
+    _Out_ JsValueRef *breakPoint)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        PARAM_NOT_NULL(breakPoint);
+
+        *breakPoint = JS_INVALID_REFERENCE;
+
+        JsrtContext *currentContext = JsrtContext::GetCurrent();
+
+        JsrtRuntime * runtime = currentContext->GetRuntime();
+
+        ThreadContextScope scope(runtime->GetThreadContext());
+
+        if (!scope.IsValid())
+        {
+            return JsErrorWrongThread;
+        }
+
+        VALIDATE_IS_DEBUGGING(runtime->GetJsrtDebugManager());
+
+        Js::Utf8SourceInfo* utf8SourceInfo = nullptr;
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && utf8SourceInfo == nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            scriptContext->GetSourceList()->MapUntil([&](int i, RecyclerWeakReference<Js::Utf8SourceInfo>* sourceInfoWeakRef) -> bool
```

Comment:
ok - you could still call scriptContext->MapSource(). That is what you are looking for.

---
In reply to: [62735056](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62735056) [](ancestors = 62735056,62733224)